# 자바 예외 이해

- 예외 계층
- 예외 기본 규칙
- 언체크 예외 기본 이해
- 체크 예외 활용
- 언체크 예외 활용
- 예외포함과 스택 트레이스

## 예외 계층

스프링이 제공하는 예외 추상화를 이해하기 위해서는 먼저 자바 기본 예외에 대한 이해가 필요하다.

![image.png](%E1%84%8C%E1%85%A1%E1%84%87%E1%85%A1%20%E1%84%8B%E1%85%A8%E1%84%8B%E1%85%AC%20%E1%84%8B%E1%85%B5%E1%84%92%E1%85%A2%2013529d746aff8038add0cb8d74fdd1d2/image.png)

- Object : 예외도 객체이다. 모든 객체의 최상위 부모는 Object 이므로, 예외의 최상위 부모도 Object 이다.
- Throwable : 최상위 예외이다. 하위에 Exception 과 Error 가 있다.
- Error : 메모리 부족이나 심각한 시스템 오류와 같이 애플리케이션에서 복구 불가능한 시스템 예외이다.
    - 애플리케이션 개발자는 이 예외를 잡으려고 해서는 안된다.
    - 상위 예외를 catch 로 잡으면, 하위 예외까지 모두 함께 잡는다. 따라서  애플리케이션 로직에서는 Throwable 예외도 잡으면 안된다. 왜? Error 예외도 함께 잡을 수 있기 때문이다.
- Exception : 체크 예외
    - 애플리케이션 로직에서 사용할 수 있는 실질적인 최상위 예외이다
    - Exception과 그 하위 예외는 모두 컴파일러가 체크하는 체크 예외이다.
    - 단, RuntimeException은 예외로 한다
- RuntimeException  : 언체크 예외, 런타임 예외
    - 컴파일러가 체크하지 않는 언체크 예외이다.
    - RuntimeException과 그 자식 예외는 모두 언체크 예외이다.
    - RuntimeException 의 이름을 따라서 RuntimeException 과 그 하위 언체크 예외를 런타임 예외라고 많이 부른다.

## 예외 기본 규칙

- 예외는 폭탄 돌리기와 같다. 잡아서 처리하고나, 처리할 수 없으면 밖으로 던져야 한다.

![image.png](%E1%84%8C%E1%85%A1%E1%84%87%E1%85%A1%20%E1%84%8B%E1%85%A8%E1%84%8B%E1%85%AC%20%E1%84%8B%E1%85%B5%E1%84%92%E1%85%A2%2013529d746aff8038add0cb8d74fdd1d2/image%201.png)

예외를 처리하지 못하면 호출한 곳으로 예외를 계속 던지게 된다.

**예외에 대해서는 2가지 기본 규칙을 기억하자!**

1. 예외는 잡아서 처리하거나 던져야 한다.
2. 예외를 잡거나 던질 때 지정한 예외 뿐만 아니라, 그 예외의 자식들도 함께 처리된다.
    - 예를 들어 Exception을 catch로 잡으면 그 하위 예외들도 모두 잡을 수 있다.
    - 예를 들어 Exception을 throws로 던지면 그 하위 예외들도 모두 던질 수 있다.

**→ 예외를 만약 처리하지 않고 계속 던지면?**

- 자바 → 메인 쓰레드에서 예외 로그 출력 후 시스템 종료 (메인 쓰레드 종료)
- WAS → spring boot 의 경우 톰켓 서버 → 필터 … 까지 올라가고, 500에러 처리됨. → 기본 혹은 별도 지정한 오류코드에 따른 반환 값, 페이지 를 반환한다.

 

## 체크 예외 기본 이해

```java
package hello.jdbc.exception.basic;
import lombok.extern.slf4j.Slf4j;
import org.junit.jupiter.api.Test;
import static org.assertj.core.api.Assertions.assertThatThrownBy;
@Slf4j
public class CheckedTest {

    @Test
    void checked_catch(){
        Service service = new Service();
        service.callCatch();
    }

    @Test
    void checked_throw() {
        Service service = new Service();

        assertThatThrownBy(() -> service.callThrow()).isInstanceOf(MyCheckedException.class);

    }

    /**
     * Exception 상속 받은 예외는 체크 예외가 된다.
     */
    static class MyCheckedException extends Exception {
        public MyCheckedException(String message) {
            super(message);
        }
    }

    /**
     * Checked 예외는
     * 예외를 잡아서 처리하거나, 던지거나 둘 중 하나를 필수로 선택해야 한다.
     */

    static class Service{
        Repository repository = new Repository();

        /**
         * 예외를 잡아서 처리하는 코드
         */
        public void callCatch() {
            try {
                repository.call();
            } catch (MyCheckedException e) {
                log.info("예외 처리, message = {}", e.getMessage(), e);
            }
        }

        public void callThrow() throws MyCheckedException {
            repository.call();
        }

    }

    static class Repository {
        public void call() throws MyCheckedException {
            throw new MyCheckedException("ex");
        }
    }
}

```

→ 예외를 잡아서 처리하는 코드, 예외를 던지는 코드 비교.

- **체크 예외의 장단점**
    
    체크 예외는 예외를 잡어서 처리할 수 없을 때, 예외를 밖으로 던지는 throws 예외를 필수로 선언해야 한다.
    
    - 장점 : 개발자가 실수로 예외를 누락하지 않도록 컴파일러를 통해 문제를 잡아주는 안전장치이다.
    - 단점 : 실제로는 개발자가 모든 체크 예외를 반드시 잡거나 던지도록 해야하기 때문에, 너무 번거로운 일이 된다. 크게 신경쓰고 싶지 않은 예외까지 모두 챙겨야 한다.

## 언체크 예외 기본 이해

- RuntimeException 과 그 하위 예외는 언체크 예외로 분류된다.
- 언체크 예외는 말 그대로 컴파일러가 예외를 체크하지 않는다는 뜻이다.
- 예외를 던지는 throws 를 선언하지 않고, 생략할 수 있다. 이 경우 자동으로 예외를 던진다.

```java
package hello.jdbc.exception.basic;

import lombok.extern.slf4j.Slf4j;
import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

@Slf4j
public class UncheckedTest {

    @Test
    void unchecked_catch() {
        Service service = new Service();
        service.callCatch();
    }

    @Test
    void unchecked_throw() {
        Service service = new Service();
        assertThatThrownBy(() -> service.callThrow()).isInstanceOf(MyUncheckedException.class);
    }

    /**
     * RuntimeException 을 상속받은 예외는 언체크 예외가 된다.
     */
    static class MyUncheckedException extends RuntimeException {
        public MyUncheckedException(String message) {
            super(message);
        }
    }

    static class Service{
        Repository repository = new Repository();

        public void callCatch() {
            try {
                repository.call();
            } catch (MyUncheckedException e) {
                log.info("예외 처리, message={}", e.getMessage(), e);
            }
        }
        public void callThrow() {
            repository.call();
        }
    }

    static class Repository {
        public void call() {
            throw new MyUncheckedException("EX");
        }
    }
}

```

- 예외를 잡거나, 별도로 throws 하지 않아도 자동으로 예외가 던져진다.

## 체크 예외 활용

**그렇다면 언제 체크 예외를 사용하고, 언제 언체크 (런타임) 예외를 사용하면 좋을까?**

**기본 원칙은 다음 2가지를 기억하자**

- 기본적으로 언체크 예외를 사용하자
- **체크 예외는 비즈니스 로직상 의도적으로 던지는 예외에만 사용하자.**
    - 이 경우 **해당 예외를 잡아서 반드시 처리해야 하는 문제일 때만** 체크 예외를 사용해야 한다.
    - 체크 예외의 예시)
        - 계좌 이체 실패 예외
        - 결제시 포인트 부족 예외
        - 로그인 ID, PW 불일치 예외
    - 물론 이 경우에도 100% 체크 예외로 만들어야 하는 것은 아니다. 다만 매우 심각한 문제가 나올 수 있는 경우, 개발자가 실수로 예외를 놓치면 안된다고 판단할 수 있다. 이 경우 체크 예외로 만들어 두면 컴파일러를 통해 놓친 예외를 인지할 수 있다.

**체크 예외의 문제점**

체크 예외는 컴파일러가 예외 누락을 체크해주기 때문에 개발자가 실수로 예외를 놓치는 것을 막아준다.

그러나 왜 체크 예외를 기본으로 사용하는 것이 문제가 될까?

**체크 예외 문제점**

![image.png](%E1%84%8C%E1%85%A1%E1%84%87%E1%85%A1%20%E1%84%8B%E1%85%A8%E1%84%8B%E1%85%AC%20%E1%84%8B%E1%85%B5%E1%84%92%E1%85%A2%2013529d746aff8038add0cb8d74fdd1d2/image%202.png)

- Repository 나 NetworkClient 에서 온 예외를 Service 에서는 처리하지 못한다.
- 그러면 Service가 해당 예외를 Controller 에 던진다.
- 그러면 Controller도 해결 하지 못하고, 예외 공통 처리할 수 있는 필터 등으로 예외를 던져서 처리한다.

→ 이렇게 해결이 불가능한 공통 예외에는 오류 코드를 남기고, 개발자가 오류를 빨리 인지할 수 있도록 메일, 알림(문자, 슬랙 등) 을 통해 전달 받아야 한다.

**체크 예외의 문제점 - 코드 예시**

```java
package hello.jdbc.exception.basic;

import org.assertj.core.api.Assertions;
import org.junit.jupiter.api.Test;

import java.net.ConnectException;
import java.sql.SQLException;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class CheckedAppTest {

    @Test
    void checked() {
        Controller controller = new Controller();
        assertThatThrownBy(() -> controller.request()).isInstanceOf(Exception.class);
    }

    static class Controller{
        Service service = new Service();
        public void request() throws SQLException, ConnectException {
            service.logic();
        }
    }

    static class Service {
        Repository repository = new Repository();
        NetworkClient networkClient = new NetworkClient();

        public void logic() throws SQLException, ConnectException {
            repository.call();
            networkClient.call();
        }
    }

    static class NetworkClient {
        public void call() throws ConnectException {
            throw new ConnectException("연결 실패");
        }
    }

    static class Repository {
        public void call() throws SQLException {
            throw new SQLException();
        }
    }

}

```

**2가지 문제**

1. 복구 불가능한 예외
2. 의존 관계에 대한 문제

1. **복구 불가능한 예외**
    
    대부분의 예외는 복구가 불가능하다.
    
    SQLException 을 예로 들면, 데이터베이스에 무언가 문제가 있어서 발생하는 예외이다. SQL 문법에 문제가 있을 수도 있고, 데이터베이스 자체에 뭔가 문제가 발생했을 수도 있다.
    
    특히 대부분의 서비스나 컨트롤러는 이런 문제를 해결할 수는 없다. 따라서 이런 문제들은 일관성 있게 공통으로 처리해야 한다. 오류 코드를 남기고 개발자가 해당 오류를 빠르게 인지하는 것이 필요하다. 서블릿 필터, 스프링 인터셉터, 스프링의 `ControllerAdvice` 를 사용하면 이런 부분을 깔끔하게 공통으로 해결할 수 있다.
    
2. **의존 관계에 대한 문제**
    
    체크 예외의 또 다른 심각한 문제는 예외에 대한 의존 관계 문제이다.
    
    서비스나 컨트롤러에서 어쩔 수 없이 `throws` 로 던지는 예외를 선언해야 한다.
    

***그러면 서비스나 컨트롤러에서 예외를 던지는 부분을 코드에 선언하는 것이 왜 문제가 될까?***

**왜? → 서비스, 컨트롤러에서 해당 예외에 의존하기 때문!**

예시)

SQLExeption → java.sql.SQLExeption 이다. 이는 구체적인 JDBC 기술을 사용하는 것.

만약 향후 서비스를 JPA 를 사용하도록 수정 개발한다면?

서비스나 컨트롤러의 예외를 던지는 모든 코드를 SQLException → JPAException 으로 고쳐야 한다.

**결과적으로 OCP, DI 를 통해 클라이언트 코드의 변경 없이 개상 구현체를 변경할 수 있다는 장점이 체크 예외 때문에 발목을 잡게 된다.**

![image.png](%E1%84%8C%E1%85%A1%E1%84%87%E1%85%A1%20%E1%84%8B%E1%85%A8%E1%84%8B%E1%85%AC%20%E1%84%8B%E1%85%B5%E1%84%92%E1%85%A2%2013529d746aff8038add0cb8d74fdd1d2/image%203.png)

**throws Exception**

```java
void method() throws SQLException, ConnectException {..} // 이것 대신
void method() throws Exception {...} //이걸 써보자.
```

→ 깔끔해진 것 같지만, SQLException, ConnectionException  뿐만 아닌 하위 타입 모든 체크 예외를 밖으로 다 던지는 꼴이다.

→ 꼭 필요한 경우가 아니라면, Exception 자체를 밖으로 던지는 것은 좋지 않은 방법이다.

## 언체크 예외 활용

이번에는 런타임 예외를 사용해보자.

```java
package hello.jdbc.exception.basic;

import org.junit.jupiter.api.Test;

import java.net.ConnectException;
import java.sql.SQLException;

import static org.assertj.core.api.Assertions.assertThatThrownBy;

public class UncheckedAppTest {

    static class RuntimeConnectionException extends RuntimeException {
        public RuntimeConnectionException(String message) {
            super(message);
        }
    }

    static class RuntimeSQLException extends RuntimeException {
        public RuntimeSQLException() {
        }

        public RuntimeSQLException(String message) {
            super(message);
        }

        public RuntimeSQLException(String message, Throwable cause) {
            super(message, cause);
        }

        public RuntimeSQLException(Throwable cause) {
            super(cause);
        }

        public RuntimeSQLException(String message, Throwable cause, boolean enableSuppression, boolean writableStackTrace) {
            super(message, cause, enableSuppression, writableStackTrace);
        }
    }

    @Test
    void unchecked() {
        Controller controller = new Controller();
        assertThatThrownBy(() -> controller.request()).isInstanceOf(RuntimeException.class);
    }

    static class Controller{
        Service service = new Service();
        public void request() {
            service.logic();
        }
    }

    static class Service {
        Repository repository = new Repository();
        NetworkClient networkClient = new NetworkClient();

        public void logic() {
            repository.call();
            networkClient.call();
        }
    }

    static class NetworkClient {
        public void call() {
            throw new RuntimeConnectionException("연결 실패");
        }
    }

    static class Repository {
        public void call() {
            try {
                runSQL();
            } catch (SQLException e) {
                throw new RuntimeSQLException(e);
            }
        }
        public void runSQL() throws SQLException {
            throw new SQLException("ex");
        }
    }

}

```

**예외 전환**

- Repository 에서 체크 예외인 SQLException 이 발생하면 런타임 예외인 RuntimeSQLException 으로 전환 해서 예외를 던진다.
    - 참고로 이때 기존 예외를 포함해주어야 예외 출력시 스택 트레이스에서 기존 예외도 함께 확인할 수 있다.
- NetworkClient는 단순히 기존 체크 예외를 RuntimeConnectionException 이라는 런타임 예외가 발생하도록 코드를 바꾸었다.

## 예외 포함과 스택 트레이스

예외를 전환할 때는 꼭!!!!!!!!!! 기존 예외를 포함해야 한다. 그렇지 않으면 스택 트레이스를 확인할 때 심각한 문제가 발생한다.

- 로그를 출력할 때 마지막 파라미터에 예외를 넣어주면 로그에 스택 트레이스를 출력할 수 있다.
    
    
    예) `log.info("message={}", "message", ex)` , 여기에서 마지막에 `ex` 를 전달하는 것을 확인할
    수 있다. 이렇게 하면 스택 트레이스에 로그를 출력할 수 있다.
    
    예) `log.info("ex", ex)` 지금 예에서는 파라미터가 없기 때문에, 예외만 파라미터에 전달하면 스택 트레이스를 로그에 출력할 수 있다.
    
- System.out 에 스택 트레이스를 출력하려면, `e.printStackTrace()` 를 사용하면 된다.
    - 실무에서는 항상 로그를 사용한다!

**기존 예외를 포함 하는 경우**

```java
static class Repository {
  public void call() {
      try {
          runSQL();
      } catch (SQLException e) {
          throw new RuntimeSQLException(e);
      }
  }
  public void runSQL() throws SQLException {
      throw new SQLException("ex");
  }
}
```

**기존 예외를 포함하지 않는 경우**

```java
public void call() {
	try {
	    runSQL();
	} catch (SQLException e) {
		throw new RuntimeSQLException(); //기존 예외(e) 제외 
	}
}
```

→ 예외를 포함하지 않아서 기존에 발생한 java.sql.SQLException 과 스택 트레이스를 확인할 수 없다. 변환한 RuntimeSQLException 부터 예외를 확인할 수 있다. 만약 실제 DB에 연동했다면 DB에서 발생한 예외 (= root cause) 를 확인할 수 없는 심각한 문제가 발생한다.

🚨 **예외를 전환할 때는 꼭!!!!!!! 기존 예외를 포함하자!** 🚨